File: src/optical_element/g_Furion_cyliner_parabolic_focus_Mirror.cpp
#include "g_Furion_cyliner_parabolic_focus_Mirror.h"

using namespace Furion_NS;

G_Furion_Cyliner_Parabolic_Focus_Mirror::G_Furion_Cyliner_Parabolic_Focus_Mirror(G_Beam* beam_in, double ds, double di, double chi, double theta, No_Surfe* surface, double r2, Grating* grating)
    : G_Oe(beam_in, ds, di, chi, theta, surface, grating)//, center(beam_in, ds, di, chi, theta, surface, r2, grating)
{
    cout << "G_Furion_Cyliner_Parabolic_Focus_Mirror 初始化" << endl;
    //run(beam_in, ds, di, chi, theta, surface, r2, grating);
}

G_Furion_Cyliner_Parabolic_Focus_Mirror::~G_Furion_Cyliner_Parabolic_Focus_Mirror()
{
    //delete center;
    //cout << "G_Furion_Cyliner_Parabolic_Focus_Mirror 析构" << endl;
}

void G_Furion_Cyliner_Parabolic_Focus_Mirror::run(G_Beam* beam_in, double ds, double di, double chi, double theta, No_Surfe* surface, double r2, Grating* grating)
{
    cout << "G_Furion_Cyliner_Parabolic_Focus_Mirror的run" << endl;
    set_center(beam_in, ds, di, chi, theta, surface, r2, grating);
    reflect(beam_in, ds, di, chi, theta);
}

void G_Furion_Cyliner_Parabolic_Focus_Mirror::intersection(std::vector<double>& T)
{
    int n = this->n;
    
    for (int i = 0; i < n; i++) {T[i] = center->T[i];}
    cneter_to_oe_p(this->X2, this->Y2, this->Z2, center->X2, center->Y2, center->Z2);
    cout << "G_Furion_Cyliner_Parabolic_Focus_Mirror的intersection" << endl;
}

void G_Furion_Cyliner_Parabolic_Focus_Mirror::normal(std::vector<double>& Nx, std::vector<double>& Ny, std::vector<double>& Nz)
{
    cneter_to_oe_v(Nx, Ny, Nz, center->Nx, center->Ny, center->Nz);
    cout << "G_Furion_Cyliner_Parabolic_Focus_Mirror的normal" << endl;
}

void G_Furion_Cyliner_Parabolic_Focus_Mirror::matrixMulti_GFCPFM(std::vector<double>& L2, std::vector<double>& M2, std::vector<double>& N2, std::vector<double>& matrix, std::vector<double>& L, std::vector<double>& M, std::vector<double>& N, double dx, double dy, int n)
{
    for (int i = 0; i < n; i++)
    {
        L2[i] = matrix[0] * L[i] + matrix[1] * (M[i] + dy) + matrix[2] * (N[i] + dx);
        M2[i] = matrix[3] * L[i] + matrix[4] * (M[i] + dy) + matrix[5] * (N[i] + dx);
        N2[i] = matrix[6] * L[i] + matrix[7] * (M[i] + dy) + matrix[8] * (N[i] + dx);
    }
}

void G_Furion_Cyliner_Parabolic_Focus_Mirror::cneter_to_oe_p(std::vector<double>& X2, std::vector<double>& Y2, std::vector<double>& Z2, std::vector<double>& X, std::vector<double>& Y, std::vector<double>& Z)
{
    int n = this->n;

    //double *OS = new double[9];
    std::vector<double> OS(9);
    f_rx.furion_rotx(this->theta, OS);

    double dx = center->r2 - 0.5 * center->p;
    double dy = center->r2 * std::sin(2 * center->theta);
    matrixMulti_GFCPFM(X2, Y2, Z2, OS, X, Y, Z, dx, dy, n);
    
    //destory_1d(OS);
}

void G_Furion_Cyliner_Parabolic_Focus_Mirror::cneter_to_oe_v(std::vector<double>& Nx, std::vector<double>& Ny, std::vector<double>& Nz, std::vector<double>& L, std::vector<double>& M, std::vector<double>& N)
{
    int n = this->n;

    //double* OS = new double[9];
    std::vector<double> OS(9);

    f_rx.furion_rotx(this->theta, OS);

    matrixMulti(Nx, Ny, Nz, OS, L, M, N, 0, n);

    //destory_1d(OS);
}

void G_Furion_Cyliner_Parabolic_Focus_Mirror::set_center(G_Beam* beam_in, double ds, double di, double chi, double theta, No_Surfe* surface, double r2, Grating* grating)
{
    cout << "G_Furion_Cyliner_Parabolic_Collimation_Mirror的set_center" << endl;
    center = new G_Cyliner_Parabolic_Focus(beam_in, ds, di, chi, theta, surface, r2, grating);
    center->run(beam_in, ds, di, chi, theta, surface, r2, grating);
}

File: src/optical_element/MATLAB/g_Furion_cyliner_parabolic_focus_Mirror.m
classdef  g_Furion_cyliner_parabolic_focus_Mirror < g_oe%追迹phase转换成角度
  properties 
     center;     
  end
methods
function obj = g_Furion_cyliner_parabolic_focus_Mirror( beam_in,ds,di,chi,theta,surface,r2,grating)   
    obj=obj@g_oe( beam_in,ds,di,chi,theta,surface,grating) ;      %输入参数给父类   
    obj.set_center(beam_in,ds,di,chi,theta,surface,r2,grating); %求出中心坐标系中的交点和法向量  
    obj.reflect(beam_in,ds,di,chi,theta);               %反射光束线
end
end

methods%在光学镜坐标系中，计算光束和光学元器件的交点
    %[X2,Y2,Z2]是交点，默认情况是平面镜的交叉点计算方法
    function [X2,Y2,Z2,T]=intersection(obj)
        
       T=obj.center.T;
       [X2,Y2,Z2]=obj.cneter_to_oe_p(obj.center.X2,obj.center.Y2,obj.center.Z2)  ;
      
    end
end
methods%在光学平面坐标系中，平面镜的法线方向
    function [Nx,Ny,Nz]=normal(obj)
        [Nx,Ny,Nz]=obj.cneter_to_oe_v(obj.center.Nx,obj.center.Ny,obj.center.Nz) ; 
    end
end
methods%光束和光学元件交点的坐标 从center坐标系转换到光学元器件坐标系
    function  [X2,Y2,Z2]=cneter_to_oe_p(obj,X,Y,Z)  
         n=obj.beam_in.n;
         OS1=([X;Y;Z+obj.center.r2-0.5*obj.center.p])+repmat([0; obj.center.r2*sin(2*obj.center.theta);0],1,n);
         OS = Furion_rotx(obj.theta)*OS1;
         X2=OS(1,:) ;Y2=OS(2,:);Z2=OS(3,:);
    end 
end
methods%光学元器件法线 从center坐标系中转换到光学元器件坐标系中
    function  [Nx,Ny,Nz]=cneter_to_oe_v(obj,L,M,N)            
         OV=(Furion_rotx(obj.theta)*[L;M;N]);
         Nx=OV(1,:) ;Ny=OV(2,:) ;Nz=OV(3,:);
    end 
end

methods%设置转换为center坐标的类
    function set_center(obj, beam_in,ds,di,chi,theta,surface,r2,grating)
          obj.center = g_cyliner_parabolic_focus( beam_in,ds,di,chi,theta,surface,r2,grating) ;
        
    end
end
end
